docs/features/interceptors.md

```diff
@@ -0,0 +1,176 @@
+# Interceptors
+
+## Summary
+[summary]: #summary
+
+*Interceptors* are an experimental compiler feature.
+
+An *interceptor* is a method which can declaratively substitute a call to an *interceptable* method with a call to itself at compile time. This substitution occurs by having the interceptor declare the source locations of the calls that it intercepts. This provides a limited facility to change the semantics of existing code by adding new code to a compilation (e.g. in a source generator).
+
+```cs
+using System;
+using System.Runtime.CompilerServices;
+
+var c = new C();
+c.InterceptableMethod(1); // (L1,C1): prints "interceptor 1"
+c.InterceptableMethod(1); // (L2,C2): prints "other interceptor 1"
+c.InterceptableMethod(1); // prints "interceptable 1"
+
+class C
+{
+    [Interceptable]
+    public void InterceptableMethod(int param)
+    {
+        Console.WriteLine($"interceptable {param}");
+    }
+}
+
+// generated code
+static class D
+{
+    [InterceptsLocation("Program.cs", line: /*L1*/, character: /*C1*/)] // refers to the call at (L1, C1)
+    public static void InterceptorMethod(this C c, int param)
+    {
+        Console.WriteLine($"interceptor {param}");
+    }
+
+    [InterceptsLocation("Program.cs", line: /*L2*/, character: /*C2*/)] // refers to the call at (L2, C2)
+    public static void OtherInterceptorMethod(this C c, int param)
+    {
+        Console.WriteLine($"other interceptor {param}");
+    }
+}
+```
+
+## Detailed design
+[design]: #detailed-design
+
+### InterceptableAttribute
+
+A method must indicate that its calls can be *intercepted* by including `[Interceptable]` on its declaration.
+
+If a call is intercepted to a method which lacks this attribute, a warning is reported, and interception still occurs. This may be changed to an error in the future.
+
+```cs
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class InterceptableAttribute : Attribute { }
+}
+```
+
+### InterceptsLocationAttribute
+
+A method indicates that it is an *interceptor* by adding one or more `[InterceptsLocation]` attributes. These attributes refer to the source locations of the calls it intercepts.
+
+```cs
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public sealed class InterceptsLocationAttribute(string filePath, int line, int character) : Attribute
+    {
+    }
+}
+```
+
+PROTOTYPE(ic): open question in https://github.com/dotnet/roslyn/pull/67432#discussion_r1147822738
+
+Are the `[InterceptsLocation]` attributes dropped when emitting the methods to metadata?
```

Comment:
Can `InterceptsLocationAttribute` appear on a PE symbol? I think in that case, it will not be honored (by design).
I'm not sure whether we should mention that in the spec, but it would be good to have a test for that scenario.